rep-0149.rst

```diff
@@ -346,8 +346,8 @@ Example
     </description>
     <maintainer email="someone@example.com">Someone</maintainer>
 
-    <license>BSD</license>
-    <license file="LICENSE">LGPL</license>
+    <license file="LICENSE">BSD-3-Clause</license>
+    <license source-files="include/my_package/linear_math/*">Zlib</license>
```

Comment:
I fully agree that disjoint sets of files should be supported. The _machine-readable debian/copyright file_ policy specifies that a `Files` field (which is the equivalent to the proposed `source-files` attribute) may consist of multiple patterns separated by whitespace, cf. https://www.debian.org/doc/packaging-manuals/copyright-format/1.0/#files-field.
Allowing the same in the source-files attribute would be not the typical XML-way, but would avoid new XML-tags and the mapping from package.xml to the machine-readable debian/copyright file would be straightforward and obvious. Would this be fine for you, @clalancette?

Regarding your example `<license>Zlib</license> - would assume to apply to all files not explicitly listed elsewhere`: That's what I intended by the sentence `If the attribute [source-files] is not specified, the tag again refers to all source files of the package`.

Reply:
> Allowing the same in the source-files attribute would be not the typical XML-way, but would avoid new XML-tags and the mapping from package.xml to the machine-readable debian/copyright file would be straightforward and obvious. Would this be fine for you, @clalancette?

It's pretty ugly, and puts a lot of data into an attribute (which isn't very XML-like).  I'd much prefer to use new XML elements; as long as the whole thing is backwards compatible, do you see an issue with adding new elements?

Reply:
> [...] and puts a lot of data into an attribute

I agree, but if somebody is used to the `style` attribute in HTML, it's not too much in my opinion 😄

Note that the package.xml is often edited in text editors. Many nested tags increase the risk of typos and reduce readability. Please compare

```
<license>
  <source-files><file>src/foo.cpp</file><file>include/mypackage/source.hpp</file></source-files>
  <identifier>Zlib</identifier>
</license>
```

with

```
<license source-files="src/foo.cpp include/mypackage/source.hpp">ZLib</license>
```

Also, the fact that backward compatibility is achieved by allowing two locations for the actual license identifier in the most common case, namely `<license>TheIdentifier</license>` and `<license><identifier>TheIdentifier<identifier></license>`, introduces an unnecessary complexity in my opinion.

Ultimately, it's a matter of taste ... I would propose to wait for further reviews.

Reply:
@clalancette suggested and dismissed what I think is the best solution, multiple tags:

```
<license>BSD-3-Clause</license>
<license source-files="src/foo.cpp">ZLib</license>
<license source-files="include/mypackage/source.hpp">ZLib</license>
```

This is clear to everybody, no harder to parse, and no harder to write (just more verbose). If a user is needing to specify license for specific disjoint sets of files already, splitting them into separate license tags doesn't seem overburdensome to me.

The downside I see is if a file is included in more than one line, especially when globbing: does it get both licenses? Either? @ralph-lange does Debian specify what happens when you do that?

Reply:
I agree that using the attribute with space separated strings is ugly and not very XML like. But it will only get very long if you have a very disjoint set of licenses. And is pragmatic. 

The [proposal](https://github.com/ros-infrastructure/rep/pull/347#discussion_r856415069) by @jatowler was where I jumped to myself too. 

This keeps the syntax backwards compatible. The debian generator can merge things with the same string value and above discussed with matching `type` and `file` contents. If they don't exactly match, make a new stanza. Our scripts should not be doing any approximation heuristics, just matching. If they don't match they get expanded. And the goal of this is to be machine readable which values precision over terseness. 

There will be very few packages with large complicated structures with many different licenses hopefully. If they do have that level of complexity and don't match the generated output will reflect the reality of the package.

Reply:
Hi @tfoote, maybe there is a misunderstanding about the case that I meant. Consider a ROS package `my_ros2_package` by John Doe under Apache 2 which includes a third-party open-source software library `cool_math_lib` by Jane Smith under 3-clause BSD. For technical reasons, John Doe decides to distribute the files of `cool_math_lib` to the `include` and `src` folders of his `my_ros2_package`.

If multiple patterns are allowed inside the `source-files` attribute, license tags would be
```
<license file="LICENSE" type="SPDX">Apache-2.0</license>
<license source-files="include/my_ros2_package/cool_math_lib/* src/my_ros2_package/cool_math_lib/*"
 file="OSS_material/cool_math_lib_license.txt" type="SPDX">BSD-3-Clause</license>
```
which gives two `Files` paragraphs in the [machine-readable debian/copyright file format]( https://www.debian.org/doc/packaging-manuals/copyright-format/1.0/#files-paragraph). One for the source files in `my_ros2_package` written by John Doe and one for the included `cool_math_lib`:
```
Files: *
Copyright: 2022 John Doe
License: Apache-2.0

Files: include/my_ros2_package/cool_math_lib/* src/my_ros2_package/cool_math_lib/*
Copyright: 2017 Jane Smith
License: BSD-3-Clause
```

If `source-files` allows only one path or pattern, the license tags would be
```
<license source-files="*" file="LICENSE" type="SPDX">Apache-2.0</license>
<license source-files="include/my_ros2_package/cool_math_lib/*"
 file="OSS_material/cool_math_lib_license.txt" type="SPDX">BSD-3-Clause</license>
<license source-files="include/my_ros2_package/cool_math_lib/*"
 file="OSS_material/cool_math_lib_license.txt" type="SPDX">BSD-3-Clause</license>
```
which implies that the license information for the `cool_math_lib` appears two times in the package.xml and (in case of a straight-forward translation) also two times inside the machine-readable debian/copyright file. I would really like to avoid such duplication - and not by some heuristic that tries to detect and merge such duplication but by allowing multiple paths/patterns in the license tag.

Reply:
@ralph-lange I agree that we shouldn't use any non-precise heuristics. However if we can get an exact match and it's "preferable" to have a more concise listing output that's relatively easy to do. If the rest of the content in the paragraph is identical then it's trivial for the generator to collapse the representation. If it's not identical it shouldn't be collapsed. That's not a heuristic that can cause missclasification.

It's a tradeoff of keeping the parsing simpler versus reducing the number of lines. In this case I would like to bias towards keeping the parsing simpler. Especially with the wildcard matches and precedent logic we will have to support repeating of the same license paragraphs in the case of embedded content, aka file foo.txt with license A inside directory bar with license B everywhere else. In a package with license A by default. And this might be a reason to not collapse at all, or only neighboring ones. 

Requiring N elements where N is the number of disjoint sets of glob matches doesn't seem to be a large burden. For something like the generated copyright file. It's not really supposed to be human readable. If we were looking for that I would recommend writing out the tree with all the filenames explicitly, instead of requiring the user to understand the glob pattern and apply it to a parallel tree. 

I'm not strongly opinionated on using the space separated syntax, if we have standard universal parsers and can write full schemas that will check them. But I worry that if we try to support the more complex formats with multiple matches, splitting strings in any language that wants to parse the xml wll get more complicated.

Reply:
I like this approach. Sound like a good compromise between simplicity/understandability vs. conciseness. I propose to proceed as follows: Before merging this PR, I'll implement a first prototype for generating a machine-readable debian/copyright file with some basic linting capabilities and will test it on a number of common packages. If this works well, we can finalize this PR.
Since I currently have a many other tasks, the prototype will need a few weeks.

Reply:
You can find the prototype of copyright file generation here: https://github.com/boschresearch/ros_license_toolkit/tree/add/copyright_generation_demo You must run the script with the `-c` flag.

Reply:
For example for rclcpp the copyright file looks like ...
```
Format: https://www.debian.org/doc/packaging-manuals/copyright-format/1.0/
Source: tbd
Upstream-Name: rclcpp

Files:
 **
Copyright: 2014 Open Source Robotics Foundation, Inc.
           2014-2020 Open Source Robotics Foundation, Inc.
           2015 Open Source Robotics Foundation, Inc.
           2015-2020 Open Source Robotics Foundation, Inc.
           2016 Open Source Robotics Foundation, Inc.
           2016-2017 Open Source Robotics Foundation, Inc.
           2017 Open Source Robotics Foundation, Inc.
           2018 Open Source Robotics Foundation, Inc.
           2018, Bosch Software Innovations GmbH.
           2019 Intel Corporation
           2019 Nobleo Technology
           2019 Open Source Robotics Foundation, Inc.
           2019-2020 Open Source Robotics Foundation, Inc.
           2020 Amazon.com, Inc. or its affiliates
           2020 Ericsson AB
           2020 Open Source Robotics Foundation, Inc.
           2020, Apex.AI Inc.
           2021 Open Source Robotics Foundation, Inc.
           2021, Apex.AI Inc.
           2022 Open Source Robotics Foundation, Inc.
           2023 Open Source Robotics Foundation, Inc.
           2023 Sony Group Corporation
           2023 iRobot Corporation
License: Apache-2.0
 
                                  Apache License
                            Version 2.0, January 2004
                         http://www.apache.org/licenses/
 
    TERMS AND CONDITIONS FOR USE, REPRODUCTION, AND DISTRIBUTION
 
    1. Definitions.
 
       "License" shall mean [...]
```

Reply:
For the discussion in the next TSC meeting on May 18/19: I strongly prefer the lightweight, "minimally-invasive" XML schema extension for two reasons: First, many developers edit the package.xml file simply in a text editor. Second, for legal reasons I would like make the generation of the copyright file as transparent and straightforward as possible and therefore minimize the number of fields or attribute values taken from the package.xml to the copyright file.

![image](https://github.com/ros-infrastructure/rep/assets/11229766/173720a4-742e-4bb9-929d-f18cead010aa)

Reply:
Sorry this is a long thread and I may have missed something, but here are my 2c:
* Using spaces as a separator in the attribute is a bad idea, it complicates things for file paths with spaces
* We should support multiple `<license>` tags if you want to have multiple directories with license overrides (even if they specify the same license). Realistically most packages should not have a huge number of overrides.

Reply:
The machine-readable debian copyright format proposes to use a `?` in case of spaces inside paths, cf. last sentence in [Section 6.9](https://www.debian.org/doc/packaging-manuals/copyright-format/1.0/#files-field). Of course, multiple `<license>` tags may be used for the case that two or more paths, but this will result in multiple [file stanzas](https://www.debian.org/doc/packaging-manuals/copyright-format/1.0/#files-stanza).

Reply:
@amacneil We already support multiple license files. And I like @ralph-lange's suggestion that we keep the translation as simple as possible. As such we use one glob per tag and the generator will push it into a separate stanza in the output. In most cases it will be all files. In the second most common case you'll have two tags with different globs. And in the case that you will only get to disjoint declarations which would require two tags with the same license when they are so disjoint that a single glob of the other type won't work. 

```
<license>Apache 2.0</license> <!-- default -->
<license source-files=include/*>BSD-3-Clause</license> <!-- some other license -->
<license source-files=include/other/*>Apache 2.0</license> <!-- third one only for a disjoint one that can't be excluded by the 2nd -->
```

The number of cases when you're going to get to more than 3 layers is relatively small. Unless you're doing things like intermixing individual files in a directory. In which case I'd suggest that you consider a different source layout anyway. And then we don't need to get into any custom parsing.

Reply:
@amacneil, in light of @tfoote's comment, are you okay with me marking this discussion as resolved?

Reply:
I've added a paragraph on the case of disjoint declarations and explained that these have to go into separate license tags (as mentioned above by @tfoote). Furthermore I added a final note on space characters in file names/paths (which I consider as a rare case). Can I mark this discussion as resolved, @amacneil, @tfoote?

Reply:
I think we are saying the same thing, so yes I'm happy marking as resolved.

Reply:
To be clear, I disagree with this sentence:

> Finally note that space characters in the source file paths/names - which should be a very rare case - have to be replaced by a ``?`` in the filename

I don't agree that this is "very rare", and it also seems unnecessary since space does not have any special meaning.

Reply:
As a consequence we would have to replace the space characters by `?` during the translation/generation of the copyright file. (If the spaces would be kept, `include/a b.cpp` would interpreted as two patterns `include/a` and `b.cpp` in the resulting copyright file according to the Machine-readable debian/copyright file format.) However, even with such automated substitution, the set of matched files could change. (E.g., `include/a?b.cpp` matches also `include/a1b.cpp`.)

Since licenses are a legal topic, I want to avoid this substitution magic and make the generation/translation process as simple and transparent as possible. Let's simply forbid space character in the `source-files` attribute values. In case developers wants to use space characters, they can but have to take care of the `?` themselves (which can easily checked by the linter).